crates/router/src/core/files.rs

```diff
@@ -0,0 +1,104 @@
+#[cfg(feature = "s3")]
+pub mod s3_utils;
+pub mod transformers;
+
+#[cfg(not(feature = "s3"))]
+pub mod fs_utils;
+
+use api_models::files;
+use error_stack::{IntoReport, ResultExt};
+
+use super::errors::{self, RouterResponse};
+use crate::{
+    consts,
+    routes::AppState,
+    services::{self, ApplicationResponse},
+    types::{api, storage},
+};
+
+pub async fn files_create_core(
+    state: &AppState,
+    merchant_account: storage::merchant_account::MerchantAccount,
+    create_file_request: api::CreateFileRequest,
+) -> RouterResponse<files::CreateFileResponse> {
+    transformers::validate_file_upload(
+        state,
+        merchant_account.clone(),
+        create_file_request.clone(),
+    )
+    .await?;
+    let file_id = common_utils::generate_id(consts::ID_LENGTH, "file");
+    let file_key = format!("{}_{}", merchant_account.merchant_id, file_id);
+    // Check if file upload should be done to connector / self
+    let (provider_file_id, file_upload_provider) =
+        transformers::upload_and_get_provider_provider_file_id(
+            state,
+            &merchant_account,
+            &create_file_request,
+            file_key.clone(),
+        )
+        .await?;
+    let file_new = storage_models::file::FileMetadataNew {
+        file_id: file_id.clone(),
+        merchant_id: merchant_account.merchant_id,
+        file_name: create_file_request.file_name,
+        file_size: create_file_request.file_size,
+        file_type: create_file_request.file_type.to_string(),
+        provider_file_id,
+        file_upload_provider: file_upload_provider.to_string(),
+        available: true,
+    };
```

Comment:
Arguably would be better to create this in the DB before the upload and then update the provider and provider_file_upload after the upload since that way we can allow retrying uploads for the same hyperswitch file_id in case of a failure.